llmc/cuda_common.h

```diff
@@ -40,6 +41,10 @@ extern cudaDeviceProp deviceProp;
 // convenience macro for calculating grid/block dimensions for kernels
 #define CEIL_DIV(M, N) (((M) + (N)-1) / (N))
 
+// short-cuts for compile-time boolean values that can be used as function arguments
+constexpr std::bool_constant<true> True;
```

Comment:
I don't love bool_constant, it looks scary and caused an issue in an earlier PR, are we not able to avoid its use? Maybe just 0/1?

Reply:
not really. The point of bool constant is that it encodes true/false in the _type_, so it can be used for function overloading/template argument deduction.
The alternative is to specify template arguments by hand, leading to
```cpp
kernel<<<block, grid>>><true>(args)
```
which is arguably much worse than bool_constant.
(or the other way round, not sure if <<<<>>>  or <> comes first).
The issue in the other PR was that we were not including the standard header where it is defined, so it wasn't really anything specific to bool_constant.

Reply:
I guess my question is why do we need template arguments for a switch

Reply:
by having it as a template argument, we make the compiler generate two specialized versions of the function that can be optimized specifically for the task at hand. The performance gains are _probably_ really minor/below the measurement noise.